qcodes_contrib_drivers/drivers/Attocube/AttocubeController.py

```diff
@@ -0,0 +1,324 @@
+                                vals=vals.Numbers(min_value=1, max_value=10000),
+                                get_parser=self._freq_parser,
+                                snapshot_get=True
+                                )
+            # axis capacitance (gettable, will fail if Attocubes are grounded with e.g. a GND cap)
+            self.add_parameter('cap_ax{}'.format(idx),
+                               label='{} axis capacitance'.format(axis),
+                               unit='nF',
+                               get_cmd=(lambda idx=idx: self._get_cap(idx)),
+                               get_parser=self._cap_parser,
+                               snapshot_get=False
+                              )
+    
+    def ask_raw(self, cmd: str) -> str:
+        """Query instrument with cmd and return response.
+        
+        Args:
+            cmd: Command to write to controller.
+
+        Returns:
+            (str): Response of Attocube controller to the query `cmd`.
+        """
+        self.device_clear()
+        response = super().ask_raw(cmd)
+        status = super().ask_raw(cmd)
+        self.check_response(status)
+        time.sleep(0.2)
+        return response
+        
+    def write_raw(self, cmd: str) -> str:
+        """Write cmd and don't wait for response.
+        
+        Args:
+            cmd: Command to write to controller.
+        """
+        self.visa_handle.write(cmd)
+        
+    def check_response(self, response: str) -> None:
+        """Raise an exception if controller responds with 'ERROR'.
+                
+        Args:
+            response: Response from controller.
+        """
+        if 'ERROR' in response:
+            raise RuntimeError(response)
+        
+    def stop(self, axis: Union[int, str]) -> None:
+        """Stops all motion along axis and then grounds the output.
+        
+        Args:
+            axis: Either axis label (str, e.g. 'y') or index (int, e.g. 2)
+        """  
+        ax, idx = self._parse_axis(axis)
+        log.info('Stopping {} axis motion.'.format(ax))
+        self.write('stop {}'.format(idx))
+        getattr(self, 'mode_ax{}'.format(idx))('gnd')
+
+    def step(self, axis: Union[int, str], steps: int) -> None:
+        """Performs a given number of Attocube steps. steps > 0 corresponds to stepu (up),
+        steps < 0 corresponds to stepd (down).
+        
+        Args:
+            axis: Either axis label (str, e.g. 'y') or index (int, e.g. 2)
+            steps: Number of steps to perform (>0 for 'u', <0 for 'd')
+        """
+        axis, idx = self._parse_axis(axis)
+        if not isinstance(steps, int):
+            raise ValueError('steps must be an integer.')
+        freq = getattr(self, 'freq_ax{}'.format(idx))()
+        log.info('Performing {} steps along axis {}.'.format(steps, axis))
+        getattr(self, 'mode_ax{}'.format(idx))('stp')
+        direc = 'u' if steps > 0 else 'd'
```

Comment:
Can we write this logic flow normally instead of a single liner for readability? As I found this function already quite hard to read. May be some comments along the way would be helpful.

Reply:
See https://github.com/QCoDeS/Qcodes_contrib_drivers/pull/38/commits/5ed9b8ad9a87e02a13ba02a34d0e04b2a5c8ddcb